### DISCUSSION TITLE
Need Help Resolving Solidity Test Failures in FundMe Contract

### DISCUSSION THREAD

**Discussion by chapman91** [Cyfrin/foundry-full-course-cu#2187] [Q&A] [answered]
I am encountering issues with some solidity tests in my `FundMe` contract and need help from the community to resolve them. 

I have written a test to verify that the owner of the contract is set correctly during deployment. However,  the test if failing, and I am not sure why. Here are the details:

### FundMe.sol Contract 

`// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

// Note: The AggregatorV3Interface might be at a different location than what was in the video!
import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
import {PriceConverter} from "./PriceConverter.sol";

error FundMe__NotOwner();

contract FundMe {
    using PriceConverter for uint256;

    mapping(address => uint256) public addressToAmountFunded;
    address[] public funders;

    // Could we make this constant?  /* hint: no! We should make it immutable! */
    address public /* immutable */ i_owner;
     
    uint256 public constant MINIMUM_USD = 5 * 10 ** 18;  // ** acts as the exponentiation operator
   
    AggregatorV3Interface private s_priceFeed;

    // deployer address
    constructor(address priceFeed) {
        s_priceFeed = AggregatorV3Interface(priceFeed);
        i_owner = msg.sender;

    }

    function fund() public payable {
        require(msg.value.getConversionRate(s_priceFeed) >= MINIMUM_USD, "You need to spend more ETH!");
        // require(PriceConverter.getConversionRate(msg.value) >= MINIMUM_USD, "You need to spend more ETH!");
        addressToAmountFunded[msg.sender] += msg.value;
        funders.push(msg.sender);
    }
    // getVersion Function
    function getVersion() public view returns (uint256) {
        return s_priceFeed.version();
    }

    modifier onlyOwner() {
        // require(msg.sender == owner);
        if (msg.sender != i_owner) revert FundMe__NotOwner();
        _;
    }

    function withdraw() public onlyOwner {
        for (uint256 funderIndex = 0; funderIndex < funders.length; funderIndex++) {
            address funder = funders[funderIndex];
            addressToAmountFunded[funder] = 0;
        }
        funders = new address[](0);
        // // transfer
        // payable(msg.sender).transfer(address(this).balance);

        // // send
        // bool sendSuccess = payable(msg.sender).send(address(this).balance);
        // require(sendSuccess, "Send failed");

        // call
        (bool callSuccess,) = payable(msg.sender).call{value: address(this).balance}("");
        require(callSuccess, "Call failed");
    }
    // Explainer from: https://solidity-by-example.org/fallback/
    // Ether is sent to contract
    //      is msg.data empty?
    //          /   \
    //         yes  no
    //         /     \
    //    receive()?  fallback()
    //     /   \
    //   yes   no
    //  /        \
    //receive()  fallback()

    fallback() external payable {
        fund();
    }

    receive() external payable {
        fund();
    }
}`

### Deploy Contract Code 
`// SPDX-License-Identifier: MIT

pragma solidity ^0.8.18;

// Librayr enhance testing and development experience within the Foundry frame work
import {Script} from "forge-std/Script.sol";
import {FundMe} from "../src/FundMe.sol";
import {HelperConfig} from "../script/HelperConfig.sol";

contract DeployFundMe is Script {
    // Simulate the broadcast of real transactions to the blockchain
    function run() external returns (FundMe) {
        // Before startBroadcast -> Not a "real" tx
        // Simulated in a real environment
        // Creates instance of the HelperConfig contract
        HelperConfig helperConfig = new HelperConfig();

        // This line is returning a struct
        address ethUsdPriceFeed = helperConfig.activeNetworkConfig();

        // After startBroadcast -> Real tx!
        vm.startBroadcast();
        // PriceFeed
        FundMe fundMe = new FundMe(ethUsdPriceFeed);
        vm.stopBroadcast();
        // return is used to return the value of FundMe
        return fundMe;
    }
}
`
### Test Code: 

`// SPDX-License-Identifier: MIT

pragma solidity ^0.8.18;

import {Test, console} from "../lib/forge-std/src/Test.sol";
import {FundMe} from "../src/FundMe.sol";
import {DeployFundMe} from "../script/DeployFundMe.s.sol";

contract FundMeTest is Test {
    // State Variable
    FundMe fundMe;

    function setUp() external {
        // creates a new instance of the 'FundMe' contract
        // deploys new contract with seperate address and storage
        // Deploy then run test - writes to the blockchain
        // us -> FundMeTest -> FundMe
        // fundMe = new FundMe();

        DeployFundMe deployFundMe = new DeployFundMe();
        fundMe = deployFundMe.run();
    }

    function testMinimumDollarIsFive() public {
        //  5 x 10**18
        assertEq(fundMe.MINIMUM_USD(), 5e18);
    }

    // Check if owners is msg.sender
    /**
     * !
     *
     *
     * ?
     *
     * todo: we need to call some method
     *
     * @param args
     *
     *
     * @throws Exception
     */

    /**
     * ! Failed with Assertion Error
     */
    function testOwnerIsMsgSender() public {
        assertEq(fundMe.i_owner(), address(this));
    }

    /**
     * ! Failed with an EvmError: Revert
     */
    function testPriceFeedVersionIsAccurate() public {
        uint256 version = fundMe.getVersion();
        // checks for equality between two values
        // two values: expected value and actual value
        assertEq(version, 4);
    }
}
`
### **Error that occurred during the running test**

![test_fail_screenshot](https://github.com/user-attachments/assets/b07687a9-d5dd-46bd-813c-65801555bec4)

Steps I Have Tried:

1. Verified that i_owner is public.
2. Ensured DeployFundMe deploys the FundMe contract correctly.
3. Checked the test setup and deployment logic.

Despite these efforts, the test still fails. Any insights or suggestions on how to resolve this would be greatly appreciated!

**Comment by EngrPips** [comment]
Hello @chapman91, Please run your test with the `-vvvv` flag and show us the output. I would prefer you copy and paste the output here. Make sure to format it properly.

**Reply by chapman91** [reply]
I am new to using github discussion. So let me find out how to properly format the error message and send it to you.

**Reply by chapman91** [reply]
Thanks for introducing this command. It is really verbose. I copied and pasted the results from using the command 
`forge test -vvvv`: 

**Compiler Warnings:**

`Compiler run successful with warnings:
Warning (6321): Unnamed return variable can remain unassigned. Add an explicit return with value to all non-reverting code paths or name the variable.
  --> script/HelperConfig.sol:36:57:
   |
36 |     function getSepoliaEthConfig() public pure returns (NetworkConfig memory) {
   |                                                         ^^^^^^^^^^^^^^^^^^^^

Warning (2072): Unused local variable.
  --> script/HelperConfig.sol:38:9:
   |
38 |         NetworkConfig memory sepoliaConfig = NetworkConfig({
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Warning (2018): Function state mutability can be restricted to view
  --> test/FundMeTest.t.sol:24:5:
   |
24 |     function testMinimumDollarIsFive() public {
   |     ^ (Relevant source part starts here and spans across multiple lines).

Warning (2018): Function state mutability can be restricted to view
  --> test/FundMeTest.t.sol:47:5:
   |
47 |     function testOwnerIsMsgSender() public {
   |     ^ (Relevant source part starts here and spans across multiple lines).
`

**Test Results:**

` Ran 3 tests for test/FundMeTest.t.sol:FundMeTest
[PASS] testMinimumDollarIsFive() (gas: 8323)
Traces:
  [8323] FundMeTest::testMinimumDollarIsFive()
    ├─ [184] FundMe::MINIMUM_USD() [staticcall]
    │   └─ ← [Return] 5000000000000000000 [5e18]
    ├─ [0] VM::assertEq(5000000000000000000 [5e18], 5000000000000000000 [5e18]) [staticcall]
    │   └─ ← [Return] 
    └─ ← [Stop] 

[FAIL. Reason: assertion failed: 0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38 != 0x7FA9385bE102ac3EAc297483Dd6233D62b3e1496] testOwnerIsMsgSender() (gas: 10650)
Traces:
  [1179851] FundMeTest::setUp()
    ├─ [635569] → new DeployFundMe@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   └─ ← [Return] 3064 bytes of code
    ├─ [505830] DeployFundMe::run()
    │   ├─ [53650] → new HelperConfig@0x104fBc016F4bb334D775a19E8A6510109AC63E00
    │   │   └─ ← [Return] 256 bytes of code
    │   ├─ [337] HelperConfig::activeNetworkConfig() [staticcall]
    │   │   └─ ← [Return] 0x0000000000000000000000000000000000000000
    │   ├─ [0] VM::startBroadcast()
    │   │   └─ ← [Return] 
    │   ├─ [383359] → new FundMe@0x90193C961A926261B756D1E5bb255e67ff9498A1
    │   │   └─ ← [Return] 1792 bytes of code
    │   ├─ [0] VM::stopBroadcast()
    │   │   └─ ← [Return] 
    │   └─ ← [Return] FundMe: [0x90193C961A926261B756D1E5bb255e67ff9498A1]
    └─ ← [Stop] 

  [10650] FundMeTest::testOwnerIsMsgSender()
    ├─ [2381] FundMe::i_owner() [staticcall]
    │   └─ ← [Return] DefaultSender: [0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38]
    ├─ [0] VM::assertEq(DefaultSender: [0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38], FundMeTest: [0x7FA9385bE102ac3EAc297483Dd6233D62b3e1496]) [staticcall]
    │   └─ ← [Revert] assertion failed: 0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38 != 0x7FA9385bE102ac3EAc297483Dd6233D62b3e1496
    └─ ← [Revert] assertion failed: 0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38 != 0x7FA9385bE102ac3EAc297483Dd6233D62b3e1496

[FAIL. Reason: EvmError: Revert] testPriceFeedVersionIsAccurate() (gas: 10206)
Traces:
  [1179851] FundMeTest::setUp()
    ├─ [635569] → new DeployFundMe@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   └─ ← [Return] 3064 bytes of code
    ├─ [505830] DeployFundMe::run()
    │   ├─ [53650] → new HelperConfig@0x104fBc016F4bb334D775a19E8A6510109AC63E00
    │   │   └─ ← [Return] 256 bytes of code
    │   ├─ [337] HelperConfig::activeNetworkConfig() [staticcall]
    │   │   └─ ← [Return] 0x0000000000000000000000000000000000000000
    │   ├─ [0] VM::startBroadcast()
    │   │   └─ ← [Return] 
    │   ├─ [383359] → new FundMe@0x90193C961A926261B756D1E5bb255e67ff9498A1
    │   │   └─ ← [Return] 1792 bytes of code
    │   ├─ [0] VM::stopBroadcast()
    │   │   └─ ← [Return] 
    │   └─ ← [Return] FundMe: [0x90193C961A926261B756D1E5bb255e67ff9498A1]
    └─ ← [Stop] 

  [10206] FundMeTest::testPriceFeedVersionIsAccurate()
    ├─ [5099] FundMe::getVersion() [staticcall]
    │   ├─ [0] 0x0000000000000000000000000000000000000000::version() [staticcall]
    │   │   └─ ← [Stop] 
    │   └─ ← [Revert] EvmError: Revert
    └─ ← [Revert] EvmError: Revert

Suite result: FAILED. 1 passed; 2 failed; 0 skipped; finished in 57.39ms (12.06ms CPU time)

Ran 1 test suite in 906.95ms (57.39ms CPU time): 1 tests passed, 2 failed, 0 skipped (3 total tests)

Failing tests:
Encountered 2 failing tests in test/FundMeTest.t.sol:FundMeTest
[FAIL. Reason: assertion failed: 0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38 != 0x7FA9385bE102ac3EAc297483Dd6233D62b3e1496] testOwnerIsMsgSender() (gas: 10650)
[FAIL. Reason: EvmError: Revert] testPriceFeedVersionIsAccurate() (gas: 10206)

Encountered a total of 2 failing tests, 1 tests succeeded`

**Comment by EngrPips** [comment, marked as answer]
Hello @chapman91, The below test is failing because the test contract is not the owner of the `FundMe` contract. You can `console.log` the address of the owner of the `FundMe` contract and also the address of the test contract and you will see that they are indeed different

```js
function testOwnerIsMsgSender() public {
    assertEq(fundMe.i_owner(), address(this));
}
```

The below test should be failing because, at a point in your contract, you were using  sepolia chain price feed address while you were running on `Anvil` chain

```js
function testPriceFeedVersionIsAccurate() public {
    uint256 version = fundMe.getVersion();
    // checks for equality between two values
    // two values: expected value and actual value
    assertEq(version, 4);
}
```

**Reply by EngrPips** [reply]
> Hello @chapman91, The below test is failing because the test contract is not the owner of the `FundMe` contract. You can `console.log` the address of the owner of the `FundMe` contract and also the address of the test contract and you will see that they are indeed different
> 
> ```js
> function testOwnerIsMsgSender() public {
>     assertEq(fundMe.i_owner(), address(this));
> }
> ```
> 
> The below test should be failing because, at a point in your contract, you were using sepolia chain price feed address while you were running on `Anvil` chain
> 
> ```js
> function testPriceFeedVersionIsAccurate() public {
>     uint256 version = fundMe.getVersion();
>     // checks for equality between two values
>     // two values: expected value and actual value
>     assertEq(version, 4);
> }
> ```

@chapman91  I think this response address your issue properly.

**Reply by chapman91** [reply]
Whew! That was a really confusing one but I got it resolved and I understand it much better now. As I progress I believe it will come full circle. Thank you!

**Reply by EngrPips** [reply]
You are welcome, keep crushing it
